### ISSUE TITLE
HasDefaultValueSql() method missing from scaffolded context

### ISSUE THREAD

**Issue by Slushnas** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#912]
### Steps to reproduce
Scaffold a table with any column that has a default value assigned.

### The issue
In the 3.0.0-rc1.final build scaffolding produces:
```
entity.Property(e => e.Type)
                    .HasColumnName("type")
                    .HasColumnType("int(11)")
                    .HasDefaultValueSql("'0'");
```
In 3.0.0-rc2.final scaffolding produces:
```
entity.Property(e => e.Type)
                    .HasColumnName("type")
                    .HasColumnType("int(11)");
```
Is this expected behavior for the new build?

### Further technical details

MySQL version: **10.1.41-MariaDB-0ubuntu0.18.04.1**
Operating system: **Ubuntu 18.04**
Pomelo.EntityFrameworkCore.MySql version: **3.0.0-rc2.final**
Microsoft.AspNetCore.App version: **3.0**

**Comment by lauxjpn** [comment]
This is expected, as long as the column default value is the default value in .NET and the column is not nullable.
If your column's default value is defined as anything but `0` for a numeric type, it should lead to a `HasDefaultValueSql()` call being generated.

**Comment by Slushnas** [comment]
What about a varchar field that has an empty string as the default value?

In 3.0.0-rc1.final this is what is returned:
```
entity.Property(e => e.Title)
                    .HasColumnName("title")
                    .HasColumnType("varchar(256)")
                    .HasDefaultValueSql("''");
```
In rc2 `.HasDefaultValueSql("''")` is omitted.

Will that cause a problem since the default value of a string in .NET is null?

**Comment by lauxjpn** [comment]
For your sample property definition, the result should be the same in `r1` and `r2`, because the column is *nullable* (no `IsRequired()` call) and therefore its default value will be set explicitly:

```c#
entity.Property(e => e.Title)
    .HasColumnName("title")
    .HasColumnType("varchar(256)")
    .HasDefaultValueSql("''");
```

For *non-nullable* string columns, we explicitly handle empty strings a the default for values in non-nullable string columns, while SQL Server does not:

https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/33474040528b331604d7690cd89e5c01afe8deba/src/EFCore.MySql/Scaffolding/Internal/MySqlDatabaseModelFactory.cs#L275-L321

If you want to test the behavior, create the following table:

```sql
CREATE TABLE `Issue912`.`TestTable` (
  `Id` INT NOT NULL AUTO_INCREMENT,
  `NullableIntWithDefault` INT NULL DEFAULT 0,
  `NullableIntWithoutDefault` INT NULL,
  `NonNullableIntWithDefault` INT NOT NULL DEFAULT 0,
  `NonNullableIntWithoutDefault` INT NOT NULL,
  `NullableStringWithDefault` VARCHAR(45) NULL DEFAULT '',
  `NullableStringWithoutDefault` VARCHAR(45) NULL,
  `NonNullableStringWithDefault` VARCHAR(45) NOT NULL DEFAULT '',
  `NonNullableStringWithoutDefault` VARCHAR(45) NOT NULL,
  PRIMARY KEY (`Id`)
);
```

Then scaffold it:

```
dotnet ef dbcontext scaffold "server=127.0.0.1;user id=root;password=;port=3306;database=Issue912" Pomelo.EntityFrameworkCore.MySql -o Models -c Issue912Context
```

The following (correct) model is being generated:

```c#
public partial class TestTable
{
    public int Id { get; set; }
    public int? NullableIntWithDefault { get; set; }
    public int? NullableIntWithoutDefault { get; set; }
    public int NonNullableIntWithDefault { get; set; }
    public int NonNullableIntWithoutDefault { get; set; }
    public string NullableStringWithDefault { get; set; }
    public string NullableStringWithoutDefault { get; set; }
    public string NonNullableStringWithDefault { get; set; }
    public string NonNullableStringWithoutDefault { get; set; }
}
```

And the following (expected) model definition is being generated:

```c#
modelBuilder.Entity<TestTable>(entity =>
{
    entity.Property(e => e.Id)
        .HasColumnType("int(11)");

    entity.Property(e => e.NonNullableIntWithDefault)
        .HasColumnType("int(11)")
        .HasDefaultValueSql("'0'");

    entity.Property(e => e.NonNullableIntWithoutDefault)
        .HasColumnType("int(11)");

    entity.Property(e => e.NonNullableStringWithDefault)
        .IsRequired()
        .HasColumnType("varchar(45)")
        .HasDefaultValueSql("''");

    entity.Property(e => e.NonNullableStringWithoutDefault)
        .IsRequired()
        .HasColumnType("varchar(45)");

    entity.Property(e => e.NullableIntWithDefault)
        .HasColumnType("int(11)")
        .HasDefaultValueSql("'0'");

    entity.Property(e => e.NullableIntWithoutDefault)
        .HasColumnType("int(11)");

    entity.Property(e => e.NullableStringWithDefault)
        .HasColumnType("varchar(45)")
        .HasDefaultValueSql("''");

    entity.Property(e => e.NullableStringWithoutDefault)
        .HasColumnType("varchar(45)");
});
```

You can then run the following code, which works as expected:

```c#
private static void Main()
{
    using var context = new Issue912Context();

    var testRow = new TestTable()
    {
        NonNullableStringWithoutDefault = string.Empty,
    };

    context.TestTable.Add(testRow);
    context.SaveChanges();

    var result = context.TestTable.ToList();

    Debug.Assert(result.Count == 1);

    Debug.Assert(result[0].NullableIntWithDefault == 0);
    Debug.Assert(result[0].NullableIntWithoutDefault == null);
    Debug.Assert(result[0].NonNullableIntWithDefault == 0);
    Debug.Assert(result[0].NonNullableIntWithoutDefault == 0);
    
    Debug.Assert(result[0].NullableStringWithDefault == string.Empty);
    Debug.Assert(result[0].NullableStringWithoutDefault == null);
    Debug.Assert(result[0].NonNullableStringWithDefault == string.Empty);
    Debug.Assert(result[0].NonNullableStringWithoutDefault == string.Empty);
}
```

So everything looks good to me. But feel free to test it thoroughly and report back to us, if you find any issues.

**Comment by lauxjpn** [comment]
@Slushnas Nice catch regarding the empty string issue! It is fixed with #915.
